libraries/push/impl/src/main/kotlin/io/element/android/libraries/push/impl/notifications/conversations/DefaultNotificationConversationService.kt

```diff
@@ -0,0 +1,100 @@
+import io.element.android.libraries.matrix.api.core.RoomId
+import io.element.android.libraries.matrix.api.core.SessionId
+import io.element.android.libraries.matrix.api.core.ThreadId
+import io.element.android.libraries.matrix.ui.media.ImageLoaderHolder
+import io.element.android.libraries.matrix.ui.media.InitialsAvatarBitmapGenerator
+import io.element.android.libraries.push.api.notifications.NotificationBitmapLoader
+import io.element.android.libraries.push.api.notifications.conversations.NotificationConversationService
+import io.element.android.libraries.push.impl.intent.IntentProvider
+import javax.inject.Inject
+
+@ContributesBinding(AppScope::class)
+class DefaultNotificationConversationService @Inject constructor(
+    @ApplicationContext private val context: Context,
+    private val intentProvider: IntentProvider,
+    private val bitmapLoader: NotificationBitmapLoader,
+    private val matrixClientProvider: MatrixClientProvider,
+    private val imageLoaderHolder: ImageLoaderHolder,
+) : NotificationConversationService {
+    override suspend fun onSendMessage(
+        sessionId: SessionId,
+        roomId: RoomId,
+        roomName: String,
+        roomIsDirect: Boolean,
+        roomAvatarUrl: String?,
+        threadId: ThreadId?,
+    ) {
+        val categories = setOfNotNull(
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) ShortcutInfo.SHORTCUT_CATEGORY_CONVERSATION else null
+        )
+
+        val client = matrixClientProvider.getOrRestore(sessionId).getOrNull() ?: return
+        val imageLoader = imageLoaderHolder.get(client)
+
+        val defaultShortcutIconSize = ShortcutManagerCompat.getIconMaxWidth(context)
+        val useDarkTheme = context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK == Configuration.UI_MODE_NIGHT_YES
+        val icon = bitmapLoader.getRoomBitmap(roomAvatarUrl, imageLoader)?.let(IconCompat::createWithBitmap)
+            ?: InitialsAvatarBitmapGenerator(useDarkTheme = useDarkTheme)
+                .generateBitmap(defaultShortcutIconSize, AvatarData(id = roomId.value, name = roomName, size = AvatarSize.RoomHeader))
+                ?.let(IconCompat::createWithAdaptiveBitmap)
+
+        val shortcutInfo = ShortcutInfoCompat.Builder(context, "$sessionId-$roomId")
+            .setShortLabel(roomName)
+            .setIcon(icon)
+            .setIntent(intentProvider.getViewRoomIntent(sessionId, roomId, threadId = null))
+            .setCategories(categories)
+            .setLongLived(true)
+            .let {
+                when (roomIsDirect) {
+                    true -> it.addCapabilityBinding("actions.intent.SEND_MESSAGE")
+                    false -> it.addCapabilityBinding("actions.intent.SEND_MESSAGE", "message.recipient.@type", listOf("Audience"))
+                }
+            }
+            .build()
+
+        ShortcutManagerCompat.pushDynamicShortcut(context, shortcutInfo)
```

Comment:
Maybe safer to protect all methods from `ShortcutManaerCompat` with `runCatchingExceptions`, as we had some issue on Element Classic regarding this.

Reply:
https://github.com/element-hq/element-x-android/pull/5180/commits/462031ab0fac813e0e3f3960333952c591263a53